File: 2023/day01/part2.py
# with open("example2", "r") as f:
with open("input", "r") as f:
    lines = [l.strip() for l in f.readlines()]
nums = ["one", "two", "three", "four", "five", "six", "seven", "eight", "nine"]


def is_num(s):
    return s.isdigit() or s in nums


res = 0
for line in lines:
    found = False
    first_num = None
    for i in range(len(line)):
        if found:
            break
        if is_num(line[i]):
            first_num = line[i]
            found = True
            break
        for num in nums:
            candidate = line[i : i + len(num)]
            if is_num(candidate):
                first_num = str(nums.index(candidate) + 1)
                found = True
                break

    found = False
    last_num = None
    for i in range(len(line) - 1, -1, -1):
        if found:
            break
        if is_num(line[i]):
            last_num = line[i]
            found = True
            break
        for num in nums:
            candidate = line[i - len(num) + 1 : i + 1]
            if is_num(candidate):
                last_num = str(nums.index(candidate) + 1)
                found = True
                break

    res += int(first_num + last_num)
print(res)

File: 2023/day01/part2.cpp
#include <bits/stdc++.h>
using namespace std;
bool charIsNum(char &c)
{
    return c >= '0' && c <= '9';
}

char stringToNum(string &s, vector<string> &nums)
{
    for (int i = 0; i < nums.size(); i++)
    {
        if (s == nums[i])
            return i + 1 + '0';
    }
    return -1;
}

int main()
{
    vector<string> nums = {"one", "two", "three", "four", "five", "six", "seven", "eight", "nine"};
    string s, a, b;
    int partTwo = 0;
    while (cin >> s)
    {
        for (int i = 0; i < s.size(); i++)
        {
            if (charIsNum(s[i]))
            {
                a = s[i];
                goto sec;
            }
            for (auto &num : nums)
            {
                string candString = s.substr(i, num.size());
                if (candString == num)
                {
                    a = stringToNum(candString, nums);
                    goto sec;
                }
            }
        }
    sec:
        for (int i = s.size() - 1; i >= 0; i--)
        {
            if (charIsNum(s[i]))
            {
                b = s[i];
                goto ut;
            }
            for (auto &num : nums)
            {
                string candString = s.substr(i, num.size());
                if (candString == num)
                {
                    b = stringToNum(candString, nums);
                    goto ut;
                }
            }
        }
    ut:
        partTwo += stoi(a + b);
    }
    cout << "Part 2: " << partTwo << endl;
}